### ISSUE TITLE
Update probabilities for stratified_sampling

### ISSUE THREAD

**Issue by jrabary** [tensorflow/tensorflow#3385]
The function `tf.contrib.framework.sampling_ops.stratified_sample` doesn't accept probability update. Is it a bug or a normal behaviour ?
Bellow is a sample code that produce the error. The idea is to resample a label sub set from large dataset with large number of label (and unbalanced) and update the target prob after each iteration : 

```
target_num_labels = 10
num_labels = 1000

p = [float(target_num_labels)/float(num_labels)]*num_labels
label_sampler = tf.contrib.distributions.Bernoulli(p=p, dtype=tf.float32)
label_samples = tf.squeeze(label_sampler.sample(1))
target_prob = tf.div(label_samples, tf.reduce_sum(label_samples))

images, labels = ....

init_prob = ... # unbalanced label distribution from the data set
[images], labels = tf.contrib.framework.stratified_sample([images], 
     labels, init_prob, target_prob, 256, enqueue_many=True)
```

**Comment by joel-shor** [comment]
A few comments:
1) `tf.contrib.distributions.Bernoulli` is not a Tensor. If it has a registered conversion to a Tensor then your code will be fine in that respect, but if not it should throw a graph-build error.
2) The tensor `target_prob` represents "Target class proportions in batch". It should be a decimal number representing the target fraction per class. For example, if you have 3 classes, a 1-D Tensor with 3 elements would be acceptable. Its values could change each run of the graph, and it could take values like [.3, .4, .3], [.2, .7, .1], etc (note that these are all valid probability distributions).

Does that answer your question?

**Comment by jrabary** [comment]
Yes  this anwer my question. Finally, I use a non trainable variable to store the target_prob. Then with the tf.assign function I update the variable every run.

**Comment by jrabary** [comment]
Hi all, 
Did some of you use the stratified sampling function for a problem where we have a large number of classes. The current implementation is relatively slow in this situation and I'm trying to figure out a way to improve the speed in this situation.

**Comment by joel-shor** [comment]
I have been able to use this Op with thousands of classes at roughly the same speed as without this Op, in certain cases. Can you copy an example of your usage?

There are 3 general sources of slowness in this op:
1) Fixed cost of setting up extra Ops and Queues (small)
2) Cost of rejecting examples (depends on the source and target distributions)
3) Queueing delays ie enqueue Ops waiting for resources, queues being full, etc 

2) is controlled the value of the tensors you are passing the Op. The relevant information are the # of classes, source distribution, and target distribution. Can you share these with me to help debug?
3) Is controlled by the parameters you pass at initialization (queue_capacity and threads_per_queue). Can you share these with me as well?

**Comment by jrabary** [comment]
@joel-shor  Here is what exactly I'm trying to do : I have about 2000 classes that are heavily unbalanced. There can be a factor of 30 between the least and the most populated classes. I need to create a batch with a given number of classes and a given number of data per class e.g 10 classes with 20 samples per class for a batch of 200 samples. In this case after each run, I begin by sampling 10 class labels and update the target probability accordingly. In the following you have a snippets of my codes : 

~~~
# sample labels and update target prob
# target_labels : number of label to select
# num_labels : total number of labels
p = [float(target_labels)/float(num_labels)]
label_sampler = tf.contrib.distributions.Bernoulli(p=p, dtype=tf.float32)
label_samples = tf.squeeze(label_sampler.sample_n(num_labels))
label_samples = tf.while_loop(lambda l: tf.less(tf.reduce_sum(l), 3),
                                      lambda l: l,
                                      [label_samples])
target_label_prob = tf.div(label_samples, tf.reduce_sum(label_samples))
       

init_prob = np.random.binomial(1, target_labels/float(num_labels), num_labels)
init_prob = init_prob/np.sum(init_prob)

# HACK : put the target prob in a non trainable variable to make update possible
target_prob = tf.get_variable('target_prob',
                                      trainable=False, shape=[num_labels],
                                      initializer=tf.constant_initializer(init_prob, dtype=tf.float32))
update_target_prob = tf.assign(target_prob, target_label_prob)
update_target_prob = tf.group(update_target_prob,
                                      tf.Print(target_prob, [target_prob, tf.reduce_sum(label_samples)],
                                               message="Update probabilty ", first_n=50))
tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, update_target_prob)

# data_prob is computed from the data set
[imgs], lbls = tf.contrib.training.stratified_sample([images], labels, target_prob, batch_size,
                                                                  init_probs=data_prob, enqueue_many=True,
                                                                  threads_per_queue=num_threads)
~~~

**Comment by joel-shor** [comment]
If you are not careful, your target distributions can be very far from the original distribution, and you can spend a lot of time rejecting samples. This can take a lot of time even if the overhead of the Op is basically 0. Can you print out some examples of your initial and target distributions so we know it isn't that?
